mesonbuild/scripts/python_info.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
+
+# ignore all lints for this file, since it is run by python2 as well
+
+# type: ignore
+# pylint: disable=deprecated-module
+
+import os.path
+import sysconfig
+import json
+import sys
+
+def get_distutils_paths(scheme=None, prefix=None):
+    import distutils.dist
+    distribution = distutils.dist.Distribution()
+    install_cmd = distribution.get_command_obj('install')
+    if prefix is not None:
+        install_cmd.prefix = prefix
+    if scheme:
+        install_cmd.select_scheme(scheme)
+    install_cmd.finalize_options()
+    return {
+        'data': install_cmd.install_data,
+        'include': os.path.dirname(install_cmd.install_headers),
+        'platlib': install_cmd.install_platlib,
+        'purelib': install_cmd.install_purelib,
+        'scripts': install_cmd.install_scripts,
+    }
+
+# On Debian derivatives, the Python interpreter shipped by the distribution uses
+# a custom install scheme, deb_system, for the system install, and changes the
+# default scheme to a custom one pointing to /usr/local and replacing
+# site-packages with dist-packages.
+# See https://github.com/mesonbuild/meson/issues/8739.
+#
+# We should be using sysconfig, but before 3.10.3, Debian only patches distutils.
+# So we may end up falling back.
+
+def get_install_paths():
+    if sys.version_info >= (3, 10):
+        scheme = sysconfig.get_default_scheme()
+    else:
+        scheme = sysconfig._get_default_scheme()
+
+    if sys.version_info >= (3, 10, 3):
+        if 'deb_system' in sysconfig.get_scheme_names():
+            scheme = 'deb_system'
```

Comment:
This continues to be wrong on virtual environments :upside_down_face:. I really don't understand why we are still requiring users to set `python.install_env` to get Meson to install to the correct paths.

I've repeatedly asked for Meson to stop mangling the paths introspected from the Python interpreter, because that's just not portable, I really don't understand the pushback against it :disappointed:

```
root@75223faf00c2:/# python3
Python 3.11.2 (main, Feb 12 2023, 00:48:52) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig, sys
>>> from pprint import pprint
>>> sysconfig.get_scheme_names()
('deb_system', 'nt', 'nt_user', 'nt_venv', 'osx_framework_user', 'posix_home', 'posix_local', 'posix_prefix', 'posix_user', 'posix_venv', 'venv')
>>> sys.path
['', '/usr/lib/python311.zip', '/usr/lib/python3.11', '/usr/lib/python3.11/lib-dynload', '/usr/local/lib/python3.11/dist-packages', '/usr/lib/python3/dist-packages']
>>> pprint(sysconfig.get_paths('deb_system'))
{'data': '/usr',
 'include': '/usr/include/python3.11',
 'platinclude': '/usr/include/python3.11',
 'platlib': '/usr/lib/python3/dist-packages',
 'platstdlib': '/usr/lib/python3.11',
 'purelib': '/usr/lib/python3/dist-packages',
 'scripts': '/usr/bin',
 'stdlib': '/usr/lib/python3.11'}
>>> pprint(sysconfig.get_paths('deb_system'))
>>>
```
```
root@75223faf00c2:/# python3 -m venv test
root@75223faf00c2:/# test/bin/python
Python 3.11.2 (main, Feb 12 2023, 00:48:52) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig, sys
>>> from pprint import pprint
>>> sysconfig.get_scheme_names()
('deb_system', 'nt', 'nt_user', 'nt_venv', 'osx_framework_user', 'posix_home', 'posix_local', 'posix_prefix', 'posix_user', 'posix_venv', 'venv')
>>> sys.path
['', '/usr/lib/python311.zip', '/usr/lib/python3.11', '/usr/lib/python3.11/lib-dynload', '//test/lib/python3.11/site-packages']
>>> pprint(sysconfig.get_paths('deb_system'))
{'data': '//test',
 'include': '/usr/include/python3.11',
 'platinclude': '/usr/include/python3.11',
 'platlib': '//test/lib/python3/dist-packages',
 'platstdlib': '//test/lib/python3.11',
 'purelib': '//test/lib/python3/dist-packages',
 'scripts': '//test/bin',
 'stdlib': '/usr/lib/python3.11'}
>>> pprint(sysconfig.get_paths())
{'data': '//test',
 'include': '/usr/include/python3.11',
 'platinclude': '/usr/include/python3.11',
 'platlib': '//test/lib/python3.11/site-packages',
 'platstdlib': '//test/lib/python3.11',
 'purelib': '//test/lib/python3.11/site-packages',
 'scripts': '//test/bin',
 'stdlib': '/usr/lib/python3.11'}
>>>
```

Reply:
At a superficial look, I agree with @FFY00, this does not seem right. However, having recently had to deal with the same problem of computing the Python installation paths in pip, I can say that the proper solution may not be easy. One complication I am thinking about in particular is the case of compiling for the system Python but with the user specifying a prefix. For the system Python, `sysconfig.get_default_scheme()` returns the system installation paths, that may or may not be adaptable for installation in a prefix. Python on macOS (either distributed by Apple or obtained with Homebrew) are a notable example of platforms that behave this way.

However, now that the introspection script is isolated in his own source file, it should be easier to test the introspection in isolation and iron out the issues.

By the way, is Meson planning to retire support for Python 2 at some point?

Reply:
> However, having recently had to deal with the same problem of computing the Python installation paths in pip, I can say that the proper solution may not be easy. One complication I am thinking about in particular is the case of compiling for the system Python but with the user specifying a prefix.

The solution is simple, you just don't. That is not portable.

Meson, as I've also said multiple times, should be looking for the Python interpreter in the target prefix, using another interpreter and just overwriting the prefix is not portable. If the Python interpreter is missing from the target prefix, it does not make sense to be installing Python modules there anyway because there's no interpreter to pick them up.

pip inherited `--prefix` from distutils, which at the time was the source of the installation layouts, so it was able to provide this feature. Right now, it is only there for backwards compatibility, and almost no one uses it. I think the people who do use it, probably most of them, effectively use it to accomplish the same as `--root`, and should really be using `--root` in the first place.
But pip is tied to a Python installation, so it will always pick the correct paths for its installation anyway, so it's not really  an  issue. Unlike Meson, which tries to override prefix, as an attempt to calculate the install paths for other installations, instead of just using the install paths from the correct installation in the first place.

I think what should be done, as I mentioned above, is looking for Python in the target prefix, and provide a `--python` argument as an escape hatch to allow users to use a specific Python interpreter.

If I wanted to install the project libraries and data on the system, but the Python modules to a virtual environment, because I don't have a Python installation in `/usr/local` or other reason, I'd just `meson setup --python=my_venv/bin/python ...` and it would work as expected, instead of having to pass `--python.install-env=venv` to get Meson to understand what I am doing and tweak it's fragile heuristics to get the correct result on this case.
It just feels like Meson has a poor understanding of Python installation layout and management, which would be honestly fine, as it's a fair bit complicated, unless we consider that I've explained this repeatedly, was ignored, and even after that Meson went to add this `python.install-env` option to try to work around its fragile heuristics for eg.

Reply:
> I really don't understand why we are still requiring users to set python.install_env to get Meson to install to the correct paths.

For backward compatibility we kept `python.install_env` to `prefix` which is probably not what most users actually want. I think we should reconsider this and change it to `auto` by default which has the behaviour you want AFAIK.

I'm not sure `venv` and `system` values are needed, isn't `auto` always the good thing to do?

The `prefix` value is still needed I think. For example a project could be built with `--prefix /opt/something` and set PYTHONPATH appropriately when running. In that case user wants to make sure the project is selfcontained in `/opt/something` and does not pollute system paths. Arguably this is the best default because it is safest even if it means it will often not be able to run the python code out of the box. Packaging systems (distro) should always set `-Dpython.install_env=system` I think.

> It just feels like Meson has a poor understanding of Python installation layout and management

To be fair, python and debian made their absolute best to make the worst experience possible.

Reply:
> Meson, as I've also said multiple times, should be looking for the Python interpreter in the target prefix

I think we are talking about two different things. With "prefix" I intend the argument to the `--prefix` option passed to `meson setup`. This directory can be empty, thus there is no Python interpreter to look at in the prefix. This is for example the case when you are building a distribution package: you want to build for the system Python but you want to install in an isolated directory that you can package for later installation. What do you intend whit "prefix"?

> pip inherited `--prefix` from distutils, which at the time was the source of the installation layouts, so it was able to provide this feature. Right now, it is only there for backwards compatibility, and almost no one uses it.

This is not technically true. Isolated build environment for building wheels with pip are populated using `pip install --prefix`. This is exactly what I have been fixing recently in pip. But I don't see how the `--prefix` option of `pip install` is at all related to what we are discussing here.

Also, you seem to imply that given a Python interpreter there is only one set of installation paths. However, there are several standard ones and each Python distribution can add custom ones:
```
>>> sysconfig.get_scheme_names()
('nt', 'nt_user', 'nt_venv', 'osx_framework_library', 'osx_framework_user', 'posix_home', 'posix_prefix', 'posix_user', 'posix_venv', 'venv')
```
On macOS, as seen above, there are the `osx_framework_library` and `osx_framework_user`, on Debian the `deb_system` one. The default scheme returned by `sysconfig.get_default_scheme()` or the paths returned by `sysconfig.get_paths()` depend on a few things (and the Python distributor can customize the criteria). How can Meson pick the right scheme? For example, as I was explaining above, on macOS the default installation scheme for the system interpreter is `osx_framework_library` which returns absolute paths, but an user compiling for the system Python that specifies `--prefix=/tmp` would be surprised to have `meson install` try to copy files in `/usr/local/lib/python3.11/site-packages`.

Reply:
> For backward compatibility we kept `python.install_env` to `prefix` which is probably not what most users actually want. I think we should reconsider this and change it to `auto` by default which has the behaviour you want AFAIK.

I don't think this should be an option in the first place, it's unsound by design IMO.

> I'm not sure `venv` and `system` values are needed, isn't `auto` always the good thing to do?

We wouldn't need different codepaths in the first place if the introspection script was correct, like I mentioned.

```python
if 'deb_system' in sysconfig.get_scheme_names():
    scheme = 'deb_system'
```

This will always set the `deb_system` scheme, even in a virtual environment, which isn't what you want. This is a mistake I made in #9288, because I forgot to consider this scenario, and then after realizing my mistake I tried to fix in #9388, but got pushback, which was pretty frustrating given it's my fault this logic is wrong in the first place :sweat:

> The `prefix` value is still needed I think. For example a project could be built with `--prefix /opt/something` and set `PYTHONPATH` appropriately when running. In that case user wants to make sure the project is self contained in `/opt/something` and does not pollute system paths.

That is a reasonable use-case, thank you for pointing it out.

I think it's a trickier case to deal with, not just for Python. The only way I think that can work without any special options is if nothing in the project needs to link against shared libraries, because RPATH needs to be set. Does Meson have any infrastructure to support this kind of installations?
If it doesn't, do we really care that much about Python here? It's not the only thing that breaks, and I am pretty sure it can be worked around by setting `--python.platlibdir` and/or `--python.purelibdir`. If you consider these standalone installations a supported use-case, which IMO I think it'd be a nice thing to support, then Meson should have a way to let the user specify that they want a standalone installation. 
If you do already have infrastructure to deal with this, great! Then we know it's a standalone installation and we should just choose a reasonable layout ourselves, instead of mangling the paths from the Python interpreter we are using. The Python modules aren't installed in a Python installation, and are instead meant to be used `PYTHONPATH`, so the layout doesn't matter, and you'd want the layout in standalone installations to be constant anyway, not having it change because you used the Python interpreter from Debain.

> To be fair, python and debian made their absolute best to make the worst experience possible.

Eh, I am naturally biased, but I don't think there was much wrong from Python's side other than just complicating things a bit from it having virtual environments :sweat_smile:

But yes, that's fair :blush:

Reply:
After some more investigation on this issue, in the case where `-Dpython.install_env=venv` we should indeed not use `scheme = 'deb_system'`, but we should still use it in other cases.

Also, in the case is_venv is True, the default value of `python.install_env` could probably be changed from `prefix` to `venv`, or warn if it is not? or raise error if it not?

Reply:
> I think we are talking about two different things. With "prefix" I intend the argument to the `--prefix` option passed to `meson setup`. This directory can be empty, thus there is no Python interpreter to look at in the prefix.

Then you want a standalone installation. See my reply above regarding that.

> This is for example the case when you are building a distribution package: you want to build for the system Python but you want to install in an isolated directory that you can package for later installation. What do you intend whit "prefix"?

No, you shouldn't be using `--prefix` for that. `--prefix` can change details of the build, be embedded in paths on the built objects, etc.

To build a distribution package, you want to set `--prefix` with the right value for where you want to install, and use [`DESTDIR`](https://gnu.org/prep/standards/html_node/DESTDIR.html) (pip's equivalent is `--root` IIRC) to place the files in a specific directory, instead of directly on the system. See https://github.com/archlinux/svntogit-community/blob/packages/libratbag/trunk/PKGBUILD for eg. `arch-meson` is a script that calls Meson with the arguments needed to install the project on the system for packaging (eg. `--prefix=/usr`), you can find its source [here](https://github.com/archlinux/svntogit-packages/blob/packages/meson/trunk/arch-meson).

> Also, you seem to imply that given a Python interpreter there is only one set of installation paths. However, there are several standard ones and each Python distribution can add custom ones:

Sorry if it read as I was implying that, it wasn't my goal.

Yes, there are multiple installation layouts. Meson should install to the default one, with the only exception being on Debian, where the install layout is patched. But, having an option to select which scheme to install to would be good.

Reply:
> After some more investigation on this issue, in the case where `-Dpython.install_env=venv` we should indeed not use `scheme = 'deb_system'`, but we should still use it in other cases.

Yes, it should be used when you are installing to the system, the default scheme should be used otherwise.

> Also, in the case is_venv is True, the default value of `python.install_env` could probably be changed from `prefix` to `venv`, or warn if it is not? or raise error if it not?

I don't know, things will break anyway, the best thing you can do is optimize for the most common case. Or remove the heuristics and fix the Python discovery and introspection, making this irrelevant :upside_down_face: